File: src/main/java/core/procedures/bit/BItShiftLeft.java
package core.procedures.bit;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMClass;

public final class BItShiftLeft extends AFn {

  public BItShiftLeft() {
    super(new FnArgsBuilder().min(2).max(2).mandatory(new Class[] {SCMClass.BitOp.class, Long.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "bit-shift-left";
  }

  @Override
  public Long apply2(Object arg1, Object arg2) {
    return ((Number) arg1).longValue() << ((Number)arg2).longValue();
  }
}

File: src/main/java/core/procedures/bit/BItShiftLeft.kt
package core.procedures.bit

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type

class BItShiftLeft : AFn<Number?, Long>(name = "bit-shift-left", isPure = true, arity = Exactly(2),
                                        mandatoryArgsTypes = arrayOf(Type.BitOp::class.java, Long::class.javaObjectType)) {

    override operator fun invoke(arg1: Number?, arg2: Number?) = arg1!!.toLong() shl arg2!!.toInt()
}
